### PR TITLE
[READY] Add MultiForm.field_order, a declarative way of reordering subforms' fields

### PR DESCRIPTION
**Pull request by frankier**

### CODE REVIEW

**Comment by frankier** [comment]
This doesn't work exactly as intended when as_table, as_ul and friends are used. It needs at least one more pass.

**Comment by frankier** [comment]
This now behaves very similarly to the development version of Django: https://docs.djangoproject.com/en/dev/ref/forms/api/#django.forms.Form.field_order

File: tests/tests/tests.py [hunk 1]
```diff
@@ -17,11 +17,19 @@
...
     UserProfileMultiForm, BadgeMultiForm, ErrorMultiForm,
     MixedForm, NeedsFileField, ManyToManyMultiForm,
-    Step2Form,
+    Step2Form, ReorderedForm
 )
 
 
-class MultiFormTest(TestCase):
+class HtmlTest(TestCase):
+    def assertHtmlEqual(self, got, want):
```
Comment (rockymeza) [inline tests/tests/tests.py]: Django's TestCase has an assertHTMLEqual, let's use that.

File: betterforms/multiform.py [hunk 2]
```diff
@@ -17,7 +19,29 @@
 
 
 @python_2_unicode_compatible
-class MultiForm(object):
+class WhitelistedBaseForm(object):
```
Comment (rockymeza) [inline betterforms/multiform.py]: big docstring right here please. I like this idea though. It just needs heavy commenting.

File: betterforms/multiform.py [hunk 3]
```diff
@@ -17,7 +19,29 @@
...
+                 label_suffix=':', *args, **kwargs):
+        self.auto_id = auto_id
+        self.prefix = prefix
+        self.error_class = error_class
+        self.label_suffix = label_suffix
+
+    def __str__(self):
+        return self.as_table()
+
+    if '__repr__' in BaseForm.__dict__:
```
Comment (rockymeza) [inline betterforms/multiform.py]: when is this not the case?

**Comment by rockymeza** [comment]
I'm a fan of this feature. I don't think I've ever actually used `as_table` and friends because the order of the fields never made sense.

There are two things that I want to bring up:

1. betterforms has this idea of [fieldsets](http://django-betterforms.readthedocs.org/en/latest/basics.html#fieldsets). We might be interested in using that.
2. I don't like the string manipulation. Would it be possible to use the actual `(form_key, field_name)` tuple? We can override the __getitem__ to check if it's a tuple.

Overall though, I like it a lot. Thanks again for your interest in multiform, @frankier!

**Comment by frankier** [comment]
1. Interoperating with BetterForm's own fieldsets might be an interesting way to take this. I'm not actually using BetterForm (the class) since as_table and as_ul aren't actually implemented and I didn't want/need fieldsets. I'm not actually sure how well something in the style of Django's as_table would work with fieldsets. I feel like personally if/when I want more control over the form output I'm more likely to go with something like crispyforms or floppyforms.

I think the way to interoperate with BetterForm might be to change WhitelistedBaseForm to DjangoStyleOutputMixin and then factor out some sort of equivalent BetterFormStyleOutputMixin from betterform then we could change MultiForm to MultiFormMixin and end up with BetterMultiForm, DjangoMultiForm, BetterModelMultiForm, DjangoModelMultiForm, but I'm not interested in implementing this myself.

2. Yes, you're right that a tuple is a bit nicer. Done.

Other comments done

File: betterforms/multiform.py [hunk 4]
```diff
@@ -43,6 +81,34 @@ def __init__(self, data=None, files=None, *args, **kwargs):
...
+        if field_order is None:
+            return
+        fields = OrderedDict()
+
+        def add_field(key, name):
+            full_name = (key, name)
+            try:
+                fields[full_name] = self.fields.pop(full_name)
+            except KeyError:  # ignore unknown fields
+                print("unknown field", full_name)
```
Comment (rockymeza) [inline betterforms/multiform.py]: The django code just passes, I don't think we need to print.

https://github.com/django/django/blob/master/django/forms/forms.py#L124

**Comment by rockymeza** [comment]
@frankier, this is very good. I like this version very much. I think we could just remove that one print call and let's merge.

Re: `field_order` versus the fieldsets concept from BetterForm. I like supporting the same featureset as Django. Then we can always add the fieldsets on in the future.

There is one last thing that I would like to bring up. There are some slight backwards incompatibilities in this changeset. Previously the `as_(table|p|ul)` methods would call `as_(table|p|ul)` on each of the child forms, now they bypass that by using the BaseForm's `as_table` over each of the fields. In the rare case that somebody has overridden their `as_table` on a child form, that custom behavior would now be ignored.

However, seeing `as_table` never really worked well anyway (as proved by #27), I think we can ignore this minor backwards incompatibility. @acatton, @gavinwahl, @pipermerriam do you all agree?

File: betterforms/multiform.py (hunk 4 above)
Comment (frankier) [inline betterforms/multiform.py]: Oops. Good catch - thanks. This was accidentally leftover from debugging.

**Pull request closed without merging**